src/composite_core.jl

```diff
@@ -0,0 +1,139 @@
+struct PrimalAdditionFailedException{P} <: Exception
+    primal::P
+    differential::Composite{P}
+    original::Exception
+end
+
+function Base.showerror(io::IO, err::PrimalAdditionFailedException{P}) where {P}
+    println(io, "Could not construct $P after addition.")
+    println(io, "This probably means no default constructor is defined.")
+    println(io, "Either define a default constructor")
+    printstyled(io, "$P(", join(propertynames(err.differential), ", "), ")", color=:blue)
+    println(io, "\nor overload")
+    printstyled(io,
+        "ChainRulesCore.construct(::Type{$P}, ::$(typeof(err.differential)))";
+        color=:blue
+    )
+    println(io, "\nor overload")
+    printstyled(io, "Base.:+(::$P, ::$(typeof(err.differential)))"; color=:blue)
+    println(io, "\nOriginal Exception:")
+    printstyled(io, err.original; color=:yellow)
+    println(io)
+end
+
+
+
+"""
+    backing(x)
+
+Accesses the backing field of a `Composite`,
+or destructures any other composite type into a `NamedTuple`.
+Identity function on `Tuple`. and `NamedTuple`s.
+
+This is an internal function used to simplify operations between `Composite`s and the
+primal types.
+"""
+backing(x::Tuple) = x
+backing(x::NamedTuple) = x
+backing(x::Composite) = getfield(x, :backing)
+
+function backing(x::T)::NamedTuple where T
+    !isstructtype(T) && throw(DomainError(T, "backing can only be use on composite types"))
+    nfields = fieldcount(T)
+    names = ntuple(ii->fieldname(T, ii), nfields)
+    types = ntuple(ii->fieldtype(T, ii), nfields)
+
+    if @generated
+        # @btime (()->ChainRulesCore.backing(Foo(1.0, 2.0)))()
+        ## 5.590 ns (1 allocation: 32 bytes)
+
+        vals = Expr(:tuple, ntuple(ii->:(getfield(x, $ii)), nfields)...)
+        return :(NamedTuple{$names, Tuple{$(types...)}}($vals))
+    else
+        vals = ntuple(ii->getfield(x, ii), nfields)
+        return NamedTuple{names, Tuple{types...}}(vals)
+    end
+end
+
+
+
+"""
+    construct(::Type{T}, fields::[NamedTuple|Tuple])
+
+Constructs an object of type `T`, with the given fields.
+Fields must be correct in name and type, and `T` must have a default constructor.
+
+This internally is called to construct structs of the primal type `T`,
```

Comment:
We have a couple internal functions here, what are your thoughts on prefacing these with an `_`?

Reply:
`construct` is part of the public API.
Users should overload it for reasons described in the docs.
So changes to it would be breaking
So not `construct`

Maybe `backing`, I am not sure yet if that will need to be overloaded similarly,
For if one wants to represent something using different fields (kind of comes up for SVD).
Just not sure how it is going to go yet.

Reply:
No stress, I think prefixing internal functions makes it easier to know at a glance what is and is not "public".